corehq/apps/sso/tasks.py

```diff
@@ -99,3 +109,55 @@ def send_idp_cert_expires_reminder_emails(num_days):
                 f"Failed to send cert reminder email for IdP {idp}: {exc!s}",
                 exc_info=True,
             )
+
+
+@periodic_task(run_every=crontab(minute=0, hour=2), acks_late=True)
+def auto_deactivate_removed_sso_users():
+    for idp in IdentityProvider.objects.filter(
+        enable_user_deactivation=True,
+        idp_type=IdentityProviderType.AZURE_AD
+    ).all():
+        # Fetch a list of users usernames that are members of the idp
+        idp_users = idp.get_all_members_of_the_idp()
+        # Fetch a list of active WebUser usernames that are members of project spaces associated with the idp
+        domains = idp.owner.get_domains()
+        web_user_in_account = set()
+        for domain in domains:
+            [web_user_in_account.add(user.username) for user in WebUser.by_domain(domain)]
+
+        # Get criteria for exempting usernames and email domains from the deactivation list
+        authenticated_domains = AuthenticatedEmailDomain.objects.filter(identity_provider=idp)
+        sso_exempt_users = UserExemptFromSingleSignOn.objects.filter(email_domain__in=authenticated_domains)
+        exempt_usernames = sso_exempt_users.values_list('username', flat=True)
+
+        if len(idp_users) == 0 and len(web_user_in_account) - len(exempt_usernames) > 3:
```

Comment:
Another option is to encapsulate `len(web_users_in_account) - len(exempt_usernames) > 3` in a little helper where 3 is still hard coded (since it seems like this number isn't used elsewhere), but the method name and docstring provides a clear idea of what the helper is checking for/why we chose 3.

Reply:
Thank you both! Made a constant and left a comment here: e515ddada9fd16c2012217992ecb67383199a095

Reply:
I offlined with Jing about this magic number some more because I'm intrigued. She linked [this thread](https://docs.google.com/document/d/1LOF1lCl3wTsEQkEEsHtZi3u4N6kvjXaLv-Lu6e9aYjo/edit) from the tech spec which was helpful, but it seems that conversation focused on the endpoint returning an empty list and not on this magic number. @biyeun you had said:
> we should consider that an errored state and hold off from automatically deactivating Web Users if the number of active Web Users (that are not SSO Exempt) is greater than 3

with your concern being incorrectly deactivating all web users for a project. I'm curious why it can't be the case that if we are in this errored state and there are _any_ active web users that are not SSO exempt, we should hold off on deactivating web users? I know it is late in the game to bring this up, but given it is a simple change, figured it is still worth asking. Thanks!

Reply:
@gherceg I was trying to think of a case where it's likely an empty list is not an errored state, but perhaps we should always treat an empty list as a potentially errored state. thoughts?

Reply:
I prefer @gherceg 's proposal. @biyeun , do you mean always treat an empty list as a potentially errored state, even if there are no active users?

Reply:
@gherceg the reason why I chose 3 in the spec, is that felt like more of an active state of using this feature rather than a testing state. I feel if there are more than 3 active SSO users then this feature is in active use, and we should be concerned about potentially deactivating all users if the list returns empty. However, in a testing state (3 users or below) we should respect an empty list because it's possible the list of users could be empty on the azure side. Thoughts? This was stated in the spec with caution around the proposal, but no one had responded during review.

Reply:
Sorry for completely missing this thread! And yeah apologies again for bringing this up in the implementation stage and not the spec review stage.

So we are trying to differentiate between an error state, a testing state, and an active state. Given that it is hard to clearly identify a testing state, I think removing it would simplify this feature and lessen the burden on documenting this magic number of 3. If our policy is simply that when an empty list is returned from an IDP, our deactivation code will not execute and we will communicate that we have skipped this to the customer, that seems more straightforward.

But again, I'm bringing this up late and don't have all of the context, so if you feel that attempting to identify this testing state is worthwhile, feel free to move forward with this as is.

Reply:
alright. @jingcheng16 let's remove the 3 user requirement and always treat an empty list as an errored state

Reply:
@biyeun and @gherceg , the threshold removed in 66b26a40d36868585cf193b5630dc43efa7fcf22